src/request.js

```diff
@@ -0,0 +1,237 @@
+  keepAlive: true,
+});
+
+let httpsAgent = new https.Agent({
+  keepAlive: true,
+});
+
+/**
+ * Return true if these headers are ones that we'll support and false if they
+ * aren't.  We only support headers where there is a single level deep mapping
+ * of string to string values
+ */
+function validateHeaders(headers) {
+  if (typeof headers !== 'object') {
+    return false;
+  }
+  for (let key of Object.keys(headers)) {
+    if (typeof headers[key] !== 'string') {
+      return false;
+    }
+  }
+  return true;
+}
+
+class RequestError extends Error {
+
+}
+
+/**
+ * Make a request against a URL.  The provided URL will be a correctly formed
+ * URL in a string.  This method will not follow any redirects.  This method
+ * will return a promise which resolves an object like {response, request}.
+ */
+async function request(url, opts = {}) {
+  assert(typeof url === 'string', 'must provide a url');
+  let urlParts = urllib.parse(url);
+  if (!opts.allowUnsafeUrls) {
+    debug('allowing unsafe urls');
+    assert(urlParts.protocol === 'https:', 'only https is supported');
+  }
+
+  let headers = opts.headers || {};
+  if (!validateHeaders(headers)) {
+    throw new RequestError('Headers are invalid');
+  }
+  headers['user-agent'] = 'cloud-mirror' + versionString;
+
+  let port;
+  try {
+    if (urlParts.port) {
+      port = parseInt(urlParts.port, 10);
+    }
+  } catch (err) {
+    throw new RequestError('Cannot parse port'); 
+  }
+
+  let method = opts.method || 'GET';
+  method = method.toUpperCase();
+
+  assert(http.METHODS.indexOf(method) !== -1);
+  
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.data, 'if supplying a stream, you must not supply data');
+  }
+
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.stream, 'if supplying a stream, you must not supply data');
+  }
+
+  switch (urlParts.protocol) {
+    case 'https:':
+      return makeRequest({
+        httpLib: https,
+        hostname: urlParts.hostname,
+        port: port || 443,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    case 'http:':
+      return makeRequest({
+        httpLib: http,
+        hostname: urlParts.hostname,
+        port: port || 80,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    default:
+      throw new RequestError('Invalid protocol: ' + urlParts.protocol);
+      break;
+  }
+  
+}
+
+module.exports = {
```

Comment:
yep, i think that's cleaner!  but for the time being, I'd rather not change this as it's not causing any issues and I'd rather not risk it.